applications/eos-cryptography-proposal.md

```diff
@@ -0,0 +1,199 @@
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| 0a. | License | MIT |
+| 0b. | Documentation | Documentation and step by step guide will be updated in [README.md](https://github.com/ZeroPass/eosio.ck/blob/master/README.md).  | 
+| 0c. | Testing Guide | [In the guide](https://github.com/ZeroPass/eosio.ck/blob/master/README.md#algorithm-testing=), we will describe how to run those tests. |
+| 0d. | Running it | We will deploy on the [Jungle 3 and CryptoKylin testnets](https://github.com/ZeroPass/eosio.ck/blob/master/README.md#testnet=), or provide Docker if OC flag (Optimized Compilation) will be required for it to run. |
+| 1. | EOSIO SDK library | To implement different ECDSA signature verification algorithms for other ECC curves, we will write example implementation for **P-384** which will include adding **sha-384** |
+| 2. | EOSIO SDK library | Example implementation for **BrainpoolP256r1** |
+| 3. | EOSIO SDK library | Example implementation for **BrainpoolP384r1** |
+| 4. | EOSIO SDK library | Example implementation for **BrainpoolP512r1** |
+
+### Milestone 5 - A cryptographic system contract
+
+- **Estimated duration:** 1 month
+- **FTE:** 3 FTE months
+- **Costs:** 20,000 USD
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| 0a. | License | MIT |
+| 0b. | Documentation | Documentation and step by step guide will be provioded in contract repository readme | 
+| 0c. | Testing Guide | We will write instructions how to run tests in the guide posted in contract repository readme |
+| 0d. | Running it | We will deploy on the Jungle 3 and CryptoKylin testnets, or provide Docker image if running eos node in OC mode (Optimized Compilation) will be required. |
+| 0e. |	Article | 	We will publish an article that explains why adding these signature algortims makes EOS much more ready to onboard goverment supported projects. We would also present why doing the same on other blockchain platforms without hard fork is close to impossible curtesy of EOSVM speed and Optimized Compilation.|
+| 1. | EOSIO Smart Contract | A cryptographic smart contract meant to be deployed by BPs, and used by many dapp developers |
+
+Milestones 1,2,3 and 4 will be part of the crypto SDK library which anyone can use in their smart contracts. We expect that running node in OC mode will be required for some cases (we'll try to omit it), this would make it necessary for the BPs to deploy it using eosio privileged permissions. We can't expect every developer to have access to BP's contract deployment and as such, it should be deployed in a form of system contract. The contract contains all described cryptographic primitives and verification actions, and allows other contracts to pass their data for verification to it. This way only one contract needs to be maintained by the BPs, and the whole proposal can become a crypto primitive for other dapps to utilize. 
```

Comment:
> "we are not talking about privileged system accounts, but maybe some sort of “OC system accounts”.

Perhaps elaborate this in your proposal, not suggesting a deployment of a "privileged" system account, but instead a OC system enabled account (if actually required).

If the implementation can be developed without OC requirement, then this can be deployed as a standard EOS contract without any system level privileged permissions which allows any contract operator to deploy their own.

Reply:
I actually had a section about it, but internally people said it adds too much confusion (felt out of place and out of scope). Also, our OC proposed solution requires a change to core node code. I am in touch with some core devs, and the advice was that; yes this proposed solution works, but since its a bit lazy, they will look into the possibility more universal deployment of OC.

Now we have to work with uncertainty and my logic leads me to just pick up the most probable outcome (according to my read of the situation). Thats why I added some reasoning but didn't want to expand it too much. And don't feel it will be productive to add it to the grant (adds confusion that might not be relevant). Frankly, how the system will handle OC contracts on a public network is out of scope for this proposal.

We will of course keep talking with different devs to ensure some OC deployment solution gets adopted (and our Milestone 5 is compatible and deployable one way or another), once we know for sure the OC flag can't be dropped in proposed cryptographic libraries.

Reply:
What are you looking for with OC?  If it is to produce SSE or SIMD instructions that is a very scary premise as we don't validate the x86-64 that is code generated by LLVM and SSE has some areas of non-determinism that we would need to ensure don't occur.

I'm not a fan of requiring OC to be able to use this technology.  If that means host functions, then I would go that route.  There is a almost 100% chance that no matter what we do with OC or EOS-VM it will still be considerably slower than native.  If you plan on backing the host-functions via openssl, they use a liberal amount of target specific intrinsics to optimize for hardware and hard written assembly to utilize as much hardware support as possible.

I don't personally have large objections to adding more host functions, this type of problem matches the reason for the system's existence.  I would say that I do get the want to move away from adding these things at the node software level, but I don't think we are there yet in terms of the technology.

Reply:
OC will always be slower than native but by how much is very much application dependent. I've seen some applications come within a few percent. In general, computational tasks with well constructed webassembly code generation tend to do the best; and some crypto functions can fall in to that category nicely.

When running OC on WAX today, you only end up spending about 10% of the main thread's time executing WASM anyways. So if picking a pessimistic number that OC is only 50% the speed of native code: that's only a 5% loss of performance on the main thread. Thus in practice I expect the OC vs native trade off to be in the ballpark of 1-3% of main thread time for current applications.

And then of course OC vs native is completely academic at this point, whereas the improvements for EOS VM JIT to EOS VM OC are very real and within reach of a release if prioritized accordingly.

Reply:
@larryk85 OC solution is just one of the proposed steps in updated docs (starting with no OC solution, then progressing to OC, then to one using host functions). The reason why we are pushing it forward is that expecting community and bps to adopt host functions that might only be used in our contracts for a while is not something we can depend on. At the end of the day, we want to build what will be adopted.

We will definitely listen to the feedback and adopt the path that will have the most chances of succeeding at getting adopted. That being said, having benchmarks to see how far EOS-VM will need to be pushed to support that as a contract is valuable in itself we think. Might offer some guide to future EOS-VM improvements.